File: PixivModelSketch.py
# -*- coding: utf-8 -*-
# pylint: disable=C1801, C0330
import demjson3

import datetime_z
from PixivImage import PixivTagData


class SketchArtist(object):
    artistId = 0
    sketchArtistId = 0
    artistName = ""
    artistAvatar = ""
    artistToken = ""
    artistBackground = ""
    posts = []
    dateFormat = None
    _tzInfo = None
    next_page = None

    def __init__(self, artist_id, page, tzInfo=None, dateFormat=None):
        self.posts = list()
        self.dateFormat = dateFormat
        self._tzInfo = tzInfo

        if page is not None:
            post_json = demjson3.decode(page)
            self.parse_artist(post_json["data"])

    def parse_artist(self, page):
        # from https://sketch.pixiv.net/api/users/@camori.json
        root = page
        if "item" in page:
            # from https://sketch.pixiv.net/api/replies/3192562692181961341.json
            root = page["item"]["user"]

        # Issue #812
        if "pixiv_user_id" in root:
            self.artistId = root["pixiv_user_id"]
        else:
            self.artistId = root["id"]

        self.sketchArtistId = root["id"]
        self.artistName = root["name"]
        self.artistToken = root["unique_name"]
        self.artistAvatar = root["icon"]["photo"]["original"]["url"]

    def parse_posts(self, page):
        post_json = demjson3.decode(page)

        links_root = post_json["_links"]
        if "next" in links_root:
            self.next_page = links_root["next"]["href"]
        else:
            self.next_page = None

        for item in post_json["data"]["items"]:
            post_id = item["id"]
            post = SketchPost(post_id, None, None, self._tzInfo, self.dateFormat)
            post.parse_post(item)
            post.artist = self
            self.posts.append(post)

    def __str__(self):
        return f"SketchArtist({self.artistId}, {self.artistName}, {self.artistToken}, {len(self.posts)})"


class SketchPost(object):
    imageId = 0
    imageTitle = ""
    imageCaption = ""
    imageTags = None
    tags = None
    imageUrls = []
    imageResizedUrls = []
    # so far only photo is supported
    imageMode = ""
    worksDate = ""
    worksDateDateTime = None
    worksUpdateDate = ""
    worksUpdateDateTime = None

    artist = None
    dateFormat = None
    _tzInfo = None

    # not supported
    originalArtist = None
    worksResolution = ""
    worksTools = ""
    jd_rtv = 0
    jd_rtc = 0
    imageCount = 0
    fromBookmark = False
    bookmark_count = -1
    image_response_count = -1

    def __init__(self, post_id, artist, page, tzInfo=None, dateFormat=None):
        self.imageUrls = list()
        self.imageResizedUrls = list()
        self.imageId = int(post_id)
        self._tzInfo = tzInfo
        self.dateFormat = dateFormat

        if page is not None:
            post_json = demjson3.decode(page)
            if artist is None:
                artist_id = post_json["data"]["item"]["user"]["id"]
                self.artist = SketchArtist(artist_id, page, tzInfo, dateFormat)
            else:
                self.artist = artist
            self.parse_post(post_json["data"]["item"])

    def parse_post(self, page):
        # post title taken from username
        self.imageTitle = page["user"]["name"]
        self.imageCaption = page["text"]
        self.imageTags = list()
        self.tags = list()
        for tag in page["tags"]:
            self.imageTags.append(tag)
            self.tags.append(PixivTagData(tag, None))

        # add R-18 tag if is_r18 = True
        if "is_r18" in page and page["is_r18"]:
            self.imageTags.append('R-18')
            self.tags.append(PixivTagData('R-18', None))

        for media in page["media"]:
            self.imageMode = media["type"]
            self.imageUrls.append(media["photo"]["original"]["url"])
            self.imageResizedUrls.append(media["photo"]["w540"]["url"])

        self.worksDateDateTime = datetime_z.parse_datetime(str(page["published_at"]))
        if self._tzInfo is not None:
            self.worksDateDateTime = self.worksDateDateTime.astimezone(self._tzInfo)
        self.worksUpdateDateTime = datetime_z.parse_datetime(str(page["updated_at"]))
        if self._tzInfo is not None:
            self.worksUpdateDateTime = self.worksUpdateDateTime.astimezone(self._tzInfo)

        tempDateFormat = self.dateFormat or "%Y-%m-%d"  # 2018-07-22, else configured in config.ini
        self.worksDate = self.worksDateDateTime.strftime(tempDateFormat)
        self.worksUpdateDate = self.worksUpdateDateTime.strftime(tempDateFormat)

    def __str__(self):
        if self.artist is not None:
            return f"SketchPost({self.artist}: {self.imageId}, {self.imageTitle}, {self.imageMode}, {self.imageUrls[0]})"
        else:
            return f"SketchPost({self.imageId}, {self.imageTitle}, {self.imageMode}, {self.imageUrls[0]})"

File: lib/model/pixiv_model_sketch.dart
/// Pixiv Sketch model.
library;

import 'dart:convert';

import '../common/datetime_z.dart' as datetime_z;
import '../common/pixiv_helper.dart' as pixiv_helper;
import 'pixiv_image.dart';

class SketchArtist {
  int artistId = 0;
  int sketchArtistId = 0;
  String artistName = '';
  String artistAvatar = '';
  String artistToken = '';
  String artistBackground = '';
  List<SketchPost> posts = [];
  String? dateFormat;
  Duration? _tzInfo;
  String? nextPage;

  SketchArtist(int artistId, String? page,
      {Duration? tzInfo, String? dateFormat}) {
    this.dateFormat = dateFormat;
    _tzInfo = tzInfo;
    if (page != null) {
      final postJson = jsonDecode(page) as Map<String, dynamic>;
      parseArtist(postJson['data'] as Map<String, dynamic>);
    } else {
      this.artistId = artistId;
    }
  }

  void parseArtist(Map<String, dynamic> page) {
    Map<String, dynamic> root = page;
    if (page.containsKey('item')) {
      root = page['item']['user'] as Map<String, dynamic>;
    }
    if (root.containsKey('pixiv_user_id')) {
      artistId = (root['pixiv_user_id'] as num).toInt();
    } else {
      artistId = (root['id'] as num).toInt();
    }
    sketchArtistId = (root['id'] as num).toInt();
    artistName = '${root['name']}';
    artistToken = '${root['unique_name']}';
    artistAvatar = '${root['icon']['photo']['original']['url']}';
  }

  void parsePosts(String page) {
    final postJson = jsonDecode(page) as Map<String, dynamic>;
    final linksRoot = postJson['_links'] as Map<String, dynamic>?;
    if (linksRoot != null && linksRoot.containsKey('next')) {
      nextPage = '${linksRoot['next']['href']}';
    } else {
      nextPage = null;
    }
    for (final item in postJson['data']['items'] as List) {
      final postId = item['id'];
      final post = SketchPost(postId is num ? postId.toInt() : int.parse('$postId'),
          this, null, tzInfo: _tzInfo, dateFormat: dateFormat);
      post.parsePost(item as Map<String, dynamic>);
      post.artist = this;
      posts.add(post);
    }
  }

  @override
  String toString() =>
      'SketchArtist($artistId, $artistName, $artistToken, ${posts.length})';
}

class SketchPost {
  int imageId;
  String imageTitle = '';
  String imageCaption = '';
  List<String>? imageTags;
  List<PixivTagData>? tags;
  List<String> imageUrls = [];
  List<String> imageResizedUrls = [];
  String imageMode = '';
  String worksDate = '';
  DateTime? worksDateDateTime;
  String worksUpdateDate = '';
  DateTime? worksUpdateDateTime;

  SketchArtist? artist;
  String? dateFormat;
  Duration? _tzInfo;

  SketchArtist? originalArtist;
  String worksResolution = '';
  String worksTools = '';
  int jd_rtv = 0;
  int jd_rtc = 0;
  int imageCount = 0;
  bool fromBookmark = false;
  int bookmark_count = -1;
  int image_response_count = -1;

  SketchPost(this.imageId, this.artist, String? page,
      {Duration? tzInfo, this.dateFormat})
      : _tzInfo = tzInfo {
    if (page != null) {
      final postJson = jsonDecode(page) as Map<String, dynamic>;
      if (artist == null) {
        final artistId =
            (postJson['data']['item']['user']['id'] as num).toInt();
        this.artist =
            SketchArtist(artistId, page, tzInfo: tzInfo, dateFormat: dateFormat);
      }
      parsePost(postJson['data']['item'] as Map<String, dynamic>);
    }
  }

  void parsePost(Map<String, dynamic> page) {
    imageTitle = '${page['user']['name']}';
    imageCaption = '${page['text'] ?? ''}';
    imageTags = [];
    tags = [];
    for (final tag in page['tags'] as List) {
      imageTags!.add('$tag');
      tags!.add(PixivTagData('$tag', null));
    }
    if (page['is_r18'] == true) {
      imageTags!.add('R-18');
      tags!.add(PixivTagData('R-18', null));
    }
    for (final media in page['media'] as List) {
      imageMode = '${media['type']}';
      imageUrls.add('${media['photo']['original']['url']}');
      imageResizedUrls.add('${media['photo']['w540']['url']}');
    }
    worksDateDateTime =
        datetime_z.parseDatetime('${page['published_at']}');
    worksUpdateDateTime =
        datetime_z.parseDatetime('${page['updated_at']}');
    if (_tzInfo != null && worksDateDateTime != null) {
      worksDateDateTime = worksDateDateTime!.add(_tzInfo!);
    }
    if (_tzInfo != null && worksUpdateDateTime != null) {
      worksUpdateDateTime = worksUpdateDateTime!.add(_tzInfo!);
    }
    final fmt = dateFormat ?? '%Y-%m-%d';
    if (worksDateDateTime != null) {
      worksDate = pixiv_helper.strftime(fmt, worksDateDateTime!);
    }
    if (worksUpdateDateTime != null) {
      worksUpdateDate =
          pixiv_helper.strftime(fmt, worksUpdateDateTime!);
    }
    imageCount = imageUrls.length;
  }

  @override
  String toString() {
    final firstUrl = imageUrls.isNotEmpty ? imageUrls.first : '';
    if (artist != null) {
      return 'SketchPost($artist: $imageId, $imageTitle, $imageMode, $firstUrl)';
    }
    return 'SketchPost($imageId, $imageTitle, $imageMode, $firstUrl)';
  }
}
